src/mongoose_ldap_worker.erl

```diff
@@ -0,0 +1,148 @@
+                   case Encrypt of
+                       tls -> ?LDAPS_PORT;
+                       starttls -> ?LDAP_PORT;
+                       _ -> ?LDAP_PORT
+                   end),
+    RootDN = eldap_utils:get_mod_opt(rootdn, Opts,
+                     fun iolist_to_binary/1,
+                     <<"">>),
+    Password = eldap_utils:get_mod_opt(password, Opts,
+                 fun iolist_to_binary/1,
+                 <<"">>),
+    ConnectInterval = eldap_utils:get_mod_opt(connect_interval, Opts,
+                                  fun(I) when is_integer(I), I>0 -> I end,
+                                  default_connect_interval()),
+    #{handle => none,
+      servers => Servers,
+      encrypt => Encrypt,
+      tls_options => TLSOptions,
+      port => Port,
+      root_dn => RootDN,
+      password => Password,
+      connect_interval => ConnectInterval}.
+
+call_eldap(Request, State) ->
+    case do_call_eldap(Request, State) of
+        {error, Reason} when Reason =:= ldap_closed;
+                             Reason =:= {gen_tcp_error, closed} ->
+            ?INFO_MSG("LDAP request failed: connection closed, reconnecting...", []),
+            eldap:close(maps:get(handle, State)),
+            NewState = connect(State#{handle := none}),
+            retry_call_eldap(Request, NewState);
+        Result ->
+            {Result, State}
+    end.
+
+connect(State = #{handle := none,
+                  servers := Servers,
+                  encrypt := Encrypt,
+                  tls_options := TLSOptions,
+                  port := Port,
+                  root_dn := RootDN,
+                  password := Password,
+                  connect_interval := ConnectInterval}) ->
+    AnonAuth = RootDN =:= <<>> andalso Password =:= <<>>,
+    SSLConfig = case Encrypt of
+                    tls -> [{ssl, true}, {sslopts, TLSOptions}];
+                    none -> [{ssl, false}]
+                end,
+    case eldap:open(Servers, [{port, Port}, {anon_auth, AnonAuth}] ++ SSLConfig) of
+        {ok, Handle} ->
+            case eldap:simple_bind(Handle, RootDN, Password) of
+                ok ->
+                    ?INFO_MSG("Connected to LDAP server", []),
+                    State#{handle := Handle};
+                Error ->
+                    ?ERROR_MSG("LDAP authentication failed: ~p", [Error]),
+                    eldap:close(Handle),
+                    erlang:send_after(ConnectInterval, self(), connect),
+                    State
+            end;
+        Error ->
+            ?ERROR_MSG("LDAP connection failed: ~p", [Error]),
+            erlang:send_after(ConnectInterval, self(), connect),
+            State
+    end.
+
+retry_call_eldap(Request, State) ->
+    Result = do_call_eldap(Request, State),
+    {Result, State}.
+
+do_call_eldap(_Request, #{handle := none}) -> {error, not_connected};
+do_call_eldap({F, Args}, #{handle := Handle}) -> apply(eldap, F, [Handle | Args]).
+
+default_connect_interval() ->
```

Comment:
I know the optimizer will get rid of this function call, but, why isn't it a macro instead?

Reply:
Why should it be? We should only resort to macros if nothing else works. This is way easier to debug and statically analyse.